### PR TITLE
Feat env-refactor

### PR DESCRIPTION
**Pull request by better629**
**Features**

- add types of env to support further. **It is still necessary to migrate the role actions of different games later to verify the complete process**. "xxx_ext_env" means the external env to integrate with game and "xx_env" means the internal env to integrate with MG.
- add dall-e / gpt-4v(both aask and ActionNode.fill) support. `AN.fill` will be used in android app agent later.
- add dall-e & gpt-4v example
- add committed code's unit-tests
    
**Feature Docs**

**Influence**

- Current added env will't influence the main process. 
- The `Environment` moved to `metagpt/environment/base_env.py` to as BaseEnv but still used as `from metagpt.environment import Environment`
- There are some js files under `metagpt/environment/mincraft_env/mineflayer` are from `https://github.com/MineDojo/Voyager/tree/main/voyager/env/mineflayer`, they are part of mincraft env. Maybe you can skip it.

**Result**

**unit-tests**
```
pytest tests/metagpt/provider/test_openai.py::test_gen_image
============================================================================== test session starts ===============================================================================
platform darwin -- Python 3.9.17, pytest-7.2.2, pluggy-1.2.0
collected 1 item

tests/metagpt/provider/test_openai.py .                                                                                                                                    [100%]

=============================================================================== 1 passed in 48.03s ===============================================================================


pytest tests/metagpt/actions/test_action_node.py::test_action_node_with_image
============================================================================== test session starts ===============================================================================
platform darwin -- Python 3.9.17, pytest-7.2.2, pluggy-1.2.0
collected 1 item

tests/metagpt/actions/test_action_node.py .                                                                                                                                [100%]

=============================================================================== 1 passed in 9.28s ================================================================================


pytest tests/metagpt/environment/*
============================================================================== test session starts ===============================================================================
collected 6 items

tests/metagpt/environment/android_env/test_android_ext_env.py .                                                                                                            [ 16%]
tests/metagpt/environment/api/test_env_api.py .                                                                                                                            [ 33%]
tests/metagpt/environment/mincraft_env/test_mincraft_ext_env.py .                                                                                                          [ 50%]
tests/metagpt/environment/stanford_town_env/test_stanford_town_ext_env.py .                                                                                                [ 66%]
tests/metagpt/environment/test_base_env.py .                                                                                                                               [ 83%]
tests/metagpt/environment/werewolf_env/test_werewolf_ext_env.py .                                                                                                          [100%]

=============================================================================== 6 passed in 3.22s ================================================================================
```

**new example**
```
(metagpt_v2) MacBook-Pro:MetaGPT xxxx$ python3 examples/dalle_gpt4v_agent.py
2024-01-30 20:36:50.592 | INFO     | metagpt.const:get_metagpt_package_root:29 - Package root set to /Users/xxxx/work/code/MetaGPT
2024-01-30 20:36:53.838 | INFO     | metagpt.roles.role:_act:386 - MaLiang(Painter): to do GenAndImproveImageAction(GenAndImproveImageAction)
A young maiden adorned with a vibrant floral crown, clasping a bouquet of multicolored blossoms, set against a lush meadow backdrop bathed in the golden light of a serene sunset.
2024-01-30 20:37:19.797 | INFO     | metagpt.utils.cost_manager:update_cost:48 - Total running cost: $0.002 | Max budget: $10.000 | Current cost: $0.002, prompt_tokens: 65, completion_tokens: 39
2024-01-30 20:37:19.797 | INFO     | __main__:run:42 - original prompt: a girl with flowers
2024-01-30 20:37:19.797 | INFO     | __main__:run:43 - refined prompt: A young maiden adorned with a vibrant floral crown, clasping a bouquet of multicolored blossoms, set against a lush meadow backdrop bathed in the golden light of a serene sunset.
As a painter named MaLiang, I can provide an artistic evaluation of the two images.

The first image features a character in a vibrant, detailed landscape with a rich array of colors and a sense of harmony between the character and the surrounding nature. The lighting is soft and evenly distributed, creating a serene atmosphere.

The second image also presents a character in a natural setting, with a focus on a softer, golden-hour light that gives the scene a warm, inviting glow. The character's expression is more subdued and contemplative, and the overall composition has a sense of depth and realism.

Both images are beautifully crafted, with their own strengths in color, composition, and mood. The choice of which is better can be subjective and dependent on the criteria used for judgment, such as color harmony, composition, emotional impact, or technical execution.

If I were to judge based on the overall impact and the harmony between the character and the environment, I might lean towards the second image due to its warm lighting and realistic portrayal. However, this is a subjective opinion, and others may prefer the first image for its vibrant colors and fantasy-like quality.

Therefore, based on my personal artistic judgment and the criteria mentioned, I would say:

True
2024-01-30 20:37:44.822 | INFO     | metagpt.utils.cost_manager:update_cost:48 - Total running cost: $0.010 | Max budget: $10.000 | Current cost: $0.008, prompt_tokens: 59, completion_tokens: 246
2024-01-30 20:37:44.824 | INFO     | __main__:run:50 - evaluate opinion: The second generated image is better than the first one: As a painter named MaLiang, I can provide an artistic evaluation of the two images.

The first image features a character in a vibrant, detailed landscape with a rich array of colors and a sense of harmony between the character and the surrounding nature. The lighting is soft and evenly distributed, creating a serene atmosphere.

The second image also presents a character in a natural setting, with a focus on a softer, golden-hour light that gives the scene a warm, inviting glow. The character's expression is more subdued and contemplative, and the overall composition has a sense of depth and realism.

Both images are beautifully crafted, with their own strengths in color, composition, and mood. The choice of which is better can be subjective and dependent on the criteria used for judgment, such as color harmony, composition, emotional impact, or technical execution.

If I were to judge based on the overall impact and the harmony between the character and the environment, I might lean towards the second image due to its warm lighting and realistic portrayal. However, this is a subjective opinion, and others may prefer the first image for its vibrant colors and fantasy-like quality.

Therefore, based on my personal artistic judgment and the criteria mentioned, I would say:

True
```

**Other**

### CODE REVIEW

File: examples/llm_hello_world.py [hunk 1]
```diff
@@ -27,6 +29,15 @@ async def main():
     if hasattr(llm, "completion"):
         logger.info(llm.completion(hello_msg))
 
+    # check llm-vision capacity if it supports
+    invoice_path = Path(__file__).parent.joinpath("..", "tests", "data", "invoices", "invoice-2.png")
+    img_base64 = encode_image(invoice_path)
+    try:
```
Comment (geekan) [inline examples/llm_hello_world.py:35]: No exception capture is needed.

File: metagpt/environment/api/env_api.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+from pydantic import BaseModel, Field
+
+
+class EnvAPIAbstract(BaseModel):
+    """api/interface summary description"""
+
+    api_name: str = Field(default="", description="the api function name or id")
+    args: set = Field(default={}, description="the api function `args` params")
+    kwargs: dict = Field(default=dict(), description="the api function `kwargs` params")
```
Comment (geekan) [inline metagpt/environment/api/env_api.py:15]: There seems to be a need for more information here, such as desc, which can be used to describe the API, which can also come from comments and explorations, and output_type is also required

Comment (geekan) [reply metagpt/environment/api/env_api.py:15]: Further, it seems that each parameter should have a corresponding comment?

File: metagpt/environment/werewolf_env/werewolf_ext_env.py [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+        self.villager_players = [p.replace(": Villager", "") for p in self.villager_players]
+        self.special_role_players = [
+            p for p in self.living_players if p not in self.werewolf_players + self.villager_players
+        ]
+
+        # init role state
+        self.roles_state = {player_name: RoleState.ALIVE for player_name in self.living_players}
+
+    @mark_as_readable
+    def init_game_setup(
```
Comment (geekan) [inline metagpt/environment/werewolf_env/werewolf_ext_env.py:143]: Should this method be exposed?

File: tests/metagpt/environment/test_base_env.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    mark_as_readable,
+    mark_as_writeable,
+)
+
+
+class ForTestEnv(Environment):
+    value: int = 0
+
+    @mark_as_readable
+    def read_api_no_parms(self):
```
Comment (geekan) [inline tests/metagpt/environment/test_base_env.py:21]: read_api_no_param may be better

File: metagpt/environment/api/env_api.py (hunk 2 above)
Comment (geekan) [reply metagpt/environment/api/env_api.py:15]: This class isn't actually bound to EnvAPIRegistry, how does it work?

File: metagpt/environment/base_env.py [hunk 5]
```diff
@@ -1,29 +1,91 @@
...
+
+
+class ExtEnv(BaseModel):
+    """External Env to intergate actual game environment"""
+
+    def _check_api_exist(self, rw_api: Optional[str] = None):
+        if not rw_api:
+            raise ValueError(f"{rw_api} not exists")
+
+    async def observe(self, env_action: Union[str, EnvAPIAbstract]):
```
Comment (geekan) [inline metagpt/environment/base_env.py:54]: before this observe function, should we provide a observe_api/get_all_available_apis method to observe?

File: metagpt/environment/werewolf_env/werewolf_ext_env.py [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+    def get_players_status(self, player_names: list[str]) -> dict[str, RoleState]:
+        roles_state = {
+            player_name: self.roles_state[player_name]
+            for player_name in player_names
+            if player_name in self.roles_state
+        }
+        return roles_state
+
+    @mark_as_writeable
+    def wolf_kill_someone(self, player_name: str):
```
Comment (geekan) [inline metagpt/environment/werewolf_env/werewolf_ext_env.py:213]: These functions all seem to be permission-related, and at the same time, the characters that werewolves kill are actually inside ENV, and different characters have different sets of available APIs. If it seems difficult to deal with the issue of permissions with the current abstraction, what is a better way to solve it?

**Review by geekan — COMMENTED** [review]
1. In Werewolf Kill, each character has a different observation space and action space, is there a more appropriate abstraction?
2. Which ENVs are now available without additional role code?
3. observe/action seems to require a pre-emptive process to get the search space (api collection).

File: metagpt/const.py [hunk 7]
```diff
@@ -129,3 +129,45 @@ def get_metagpt_root():
...
 COMPOSITION = "Composite"
 AGGREGATION = "Aggregate"
+
+# For Android Assistant Agent
+ADB_EXEC_FAIL = "FAILED"
+
+# For Mincraft Game Agent
+MC_CKPT_DIR = METAGPT_ROOT / "data/mincraft/ckpt"
+MC_LOG_DIR = METAGPT_ROOT / "logs"
+MC_DEFAULT_WARMUP = {
```
Comment (stellaHSR) [inline metagpt/const.py:139]: 不放mg的const，放mc环境的const更合适

File: examples/llm_hello_world.py (hunk 1 above)
Comment (better629) [reply examples/llm_hello_world.py:35]: Due to we don't know which llm will be configured by the user which may don't have the image chat ability, so add a try here. Like someone will try glm-4 first.

File: metagpt/environment/werewolf_env/werewolf_ext_env.py (hunk 3 above)
Comment (better629) [reply metagpt/environment/werewolf_env/werewolf_ext_env.py:143]: It provides game_setup and role instances from the env, so mark it as a readable one.

File: tests/metagpt/environment/test_base_env.py (hunk 4 above)
Comment (better629) [reply tests/metagpt/environment/test_base_env.py:21]: spelling mistake

File: examples/llm_hello_world.py (hunk 1 above)
Comment (geekan) [reply examples/llm_hello_world.py:35]: If the LLM doesn't have this capability, then the exception should be thrown to the user

File: metagpt/environment/werewolf_env/werewolf_ext_env.py (hunk 6 above)
Comment (better629) [reply metagpt/environment/werewolf_env/werewolf_ext_env.py]: add permission check when operation and update player state after voting or hunting.

**Comment by better629** [comment]
> 1. In Werewolf Kill, each character has a different observation space and action space, is there a more appropriate abstraction?
> 2. Which ENVs are now available without additional role code?
> 3. observe/action seems to require a pre-emptive process to get the search space (api collection).

@geekan 
1. Currently, use different api with permission control. But, in truth, the werewolf can be abstracted like gym's action/space further. For now, it uses apis.
2. Android Env needs a simulator; StanfordTown can run with the_ville under `tests/data`; Mincraft needs a server; Werewolf can run local.  You can see the unittests.
3. added

File: metagpt/environment/api/env_api.py (hunk 2 above)
Comment (better629) [reply metagpt/environment/api/env_api.py:15]: > This class isn't actually bound to EnvAPIRegistry, how does it work?

EnvAPIAbstract usage can see the `tests/metagpt/environment/test_base_env.py`

File: metagpt/const.py (hunk 7 above)
Comment (better629) [reply metagpt/const.py]: `/minecraft/metagpt/minecraft_team.py` in `metagpt/environment/mincraft_env/mincraft_env.py`

**Comment by geekan** [comment]
Since there are still a lot of functions that have not been implemented and cannot be fully run, I hope to provide a readme here in a suitable place to record the work of todo, as well as the general implementation method, so that the implementation can be guided later

**Review by geekan — APPROVED** [review]
lgtm

**Pull request merged**